rust/src/wasm_helpers.rs

```diff
@@ -0,0 +1,201 @@
+
+    println!("(info, direct, exit) >");
+}
+
+// Loads and configures a WASM module from a byte array.
+pub fn load_wasm_bundle(mqtt_client: Client) -> Result<WasmInfo, Box<dyn Error>> {
+    // WASM compiler, or runtime here
+    let wasm_compiler = Cranelift::new();
+    let wasm_store = Store::new(&Universal::new(wasm_compiler).engine());
+
+    // for imported memory, exported memory is not supported in this example
+    let mem_type = MemoryType::new(5, None, false);
+    let wasm_memory = Memory::new(&wasm_store, mem_type)?;
+
+    let wasm_exists = Path::new(&CONFIGS.eea_bundle_path).exists();
+
+    if !wasm_exists {
+        println!("No EEA WASM bundle file detected at configured location.");
+
+        // create a temporary empty instance
+        let module = Module::new(&wasm_store, "(module)")?;
+        let imports = imports!{};
+        let instance = Instance::new(&module, &imports)?;
+        return Ok(WasmInfo { // defaults for no bundle
+            instance,
+            memory: wasm_memory,
+            bundle_id: DEFAULT_BUNDLE_ID.to_owned(),
+        });
+    }
+
+    // read from WASM file, if gzipped (eea_bundle_gzip set within resources/eea_config.toml) attempt to decode
+    let wasm_bytes = read(&CONFIGS.eea_bundle_path)?;
+    let module: Module = if CONFIGS.eea_bundle_gzip {
+        let mut decoder = Decoder::new(&wasm_bytes[..])?;
+        let mut decoded_data = Vec::new();
+        decoder.read_to_end(&mut decoded_data)?;
+        Module::new(&wasm_store, decoded_data)?
+    } else {
+        Module::new(&wasm_store, wasm_bytes)?
+    };
+
+    let env = Env {
+        memory: wasm_memory.clone(),
+        mqtt_client
+    };
+
+    // functions that are defined in native code and imported into the EEA WASM module
+    // user created registered functions also go here
+    let imports = imports!{
+        "env" => {
+            "memory" => wasm_memory.clone(),
+            "eea_trace" => Function::new_native_with_env(&wasm_store, env.clone(), eea_api::eea_trace),
+            "eea_send_message" => Function::new_native_with_env(&wasm_store, env.clone(), eea_api::eea_send_message),
+            "eea_get_time" => Function::new_native_with_env(&wasm_store, env.clone(), eea_api::eea_get_time),
+            "eea_get_device_id" => Function::new_native_with_env(&wasm_store, env.clone(), eea_api::eea_get_device_id),
+            "eea_storage_save" => Function::new_native_with_env(&wasm_store, env.clone(), eea_api::eea_storage_save),
+            "eea_storage_read" => Function::new_native_with_env(&wasm_store, env.clone(), eea_api::eea_storage_read),
+            "eea_set_message_buffers" => Function::new_native(&wasm_store, eea_api::eea_set_message_buffers),
+            "eea_sleep" => Function::new_native(&wasm_store, eea_api::eea_sleep),
+            // add registered_functions
+            "eea_fn_terminal_print" => Function::new_native_with_env(&wasm_store, env, registered_functions::eea_fn_terminal_print),
```

Comment:
missing `.clone` on `env`. Get a "moved" compile error when attempted to add more registered functions.

Reply:
really shouldn't clone the last one, will get a warning. I could add a note (also to the readme) for users to update this when adding a new function.